src/os/linux/mod.rs

```diff
@@ -0,0 +1,118 @@
+mod xdg;
+
+use std::{
+    ffi::OsStr,
+    fmt::{self, Display},
+    path::Path,
+};
+
+#[derive(Debug)]
+pub enum DetectEditorError {
+    LookupFailed,
+}
+
+impl Display for DetectEditorError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::LookupFailed => write!(f, "Lookup failed"),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum OpenFileError {
+    LaunchFailed,
+}
+
+impl Display for OpenFileError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::LaunchFailed => write!(f, "Launch failed"),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct Application {
+    exec_command: String,
+}
+
+impl Application {
+    pub fn detect_editor() -> Result<Self, DetectEditorError> {
+        let entry = {
+            // Try a rust code editor, then a plain text editor, if neither are available,
+            // then return an error.
+            if let Some(rust_entry) = xdg::query_mime_entry("text/rust") {
+                rust_entry
+            } else if let Some(plain_entry) = xdg::query_mime_entry("text/plain") {
+                plain_entry
+            } else {
+                return Err(DetectEditorError::LookupFailed);
+            }
+        };
+
+        for dir in xdg::get_xdg_data_dirs() {
+            // Look at the applications folder of that dir
+            let dir = dir.join("applications");
+            if let Some(result) = xdg::find_entry_in_dir(&dir, &entry) {
+                return Ok(Self {
+                    exec_command: xdg::command_from_freedesktop_entry(&result)
+                        .ok_or(DetectEditorError::LookupFailed)?
+                })
+            }
+        }
+
+        Err(DetectEditorError::LookupFailed)
+    }
+
+    pub fn open_file(&self, path: impl AsRef<Path>) -> Result<(), OpenFileError> {
+        let path = path.as_ref();
+        let path_str = path.to_string_lossy(); // Ugh, lossy
+
+        let command = xdg::build_command(&self.exec_command, &path_str)
+            .ok_or(OpenFileError::LaunchFailed)?;
+
+        // I'm having a problem creating a dettached process with bossy,
+        // but this should work and be well supported on linux systems,
+        // as "sh" and no "nohup" are pretty standard.
```

Comment:
I did some research and added functionality for this to bossy v0.1.4, so now you can just do:
```rust
bossy::Command::impure(command)
    .run_and_detach()
```
`run_and_detach` takes care of setting stdin/stdout/stderr to null, and uses `fork` and `setsid` under the hood to detach the process (inspired by [how Alacritty does it](https://github.com/alacritty/alacritty/blob/8bd2c13490f8cb6ad6b0c1104f9586b3554efea2/alacritty/src/daemon.rs)).

Since you don't need `format` anymore, `command` doesn't need to be a string. If you use [`OsStrExt::as_bytes`](https://doc.rust-lang.org/std/os/unix/ffi/trait.OsStrExt.html#tymethod.as_bytes), you can use the regex crate's [byte API](https://docs.rs/regex/1.4.2/regex/bytes/index.html) inside of `build_command` to avoid needing to use `to_string_lossy` above. Let me know if the `once_cell_regex` API doesn't play nice with that.

Reply:
You understand correctly, but I've just released once-cell-regex v0.2.1 with `once_cell_regex::byte_regex!` to cover this use-case.